### PR TITLE
US012 add new job

### PR DESCRIPTION
**Pull request by Joseph-Magee**
As an Admin

I want to add a new role to an existing capability and band
So that I can keep the software information up to date

Acceptance criteria

Able to add a new role to an existing band and capability

Able to define band for new role

Able to define job family for new role

Able to add details to new role:

Job role name

Job spec summary

Sharepoint link

Screenshot 2023-12-22 at 15 24 08

### CODE REVIEW

**Review by Criofann — APPROVED** [review]
Looks good, all checks passed and tested well

File: controller/roleController.ts [hunk 1]
```diff
@@ -13,4 +16,27 @@ module.exports = function(app: Application){
...
+        delete data.jobFamily;
+        const role: RoleFamily = data;
+        try{
+             await roleService.createRole(role);
+             await familyService.createFamily(family);
+            
+
+        }catch(e) {
+            
+
```
Comment (jkelso662) [inline controller/roleController.ts:35]: Lots of blank space!

File: service/roleService.ts [hunk 2]
```diff
@@ -10,4 +11,25 @@ module.exports.getAllRoles = async function() {
...
+};
+module.exports.createRole = async function(role: Role): Promise<number> {
+    const error: string = roleValidator.validateRole(role);
+        if (error){
+            throw new Error(error);
+        }
+        
+        try{
+            console.log("role");
+            console.log(role);
```
Comment (jkelso662) [inline service/roleService.ts:23]: I think you forgot to remove these!

File: service/roleService.ts [hunk 3]
```diff
@@ -10,4 +11,25 @@ module.exports.getAllRoles = async function() {
...
+            const response = await axios.post("http://localhost:8080/api/job-roles/", role);
+            return response.data;
+        } catch (e) {
+            if (e.response.status == 400) {
+                console.log("here");
```
Comment (jkelso662) [inline service/roleService.ts:28]: And this one :P

File: test/unit/service/familyServiceTest.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+      try {
+        await FamilyService.createFamily(JobFamily);
+      } catch (e) {
+        error = e.message;
+      }
+
+    expect(error).to.deep.equal("Invalid data");
+  });
+  it("should return Could not create family from 500 error", async () => {
```
Comment (jkelso662) [inline test/unit/service/familyServiceTest.ts:39]: Spacing/tabulation in this file looks a little off.

**Review by jkelso662 — CHANGES_REQUESTED** [review]
Very good! Just a few things to watch out for i.e. leaving in console.logs. Also would be nice if you could update your PR title to have US instead of Us. Just looks so much more consistent 😄

**Review by jkelso662 — CHANGES_REQUESTED** [review]
Code looks good! Would approve but I'll save you the time this getting to signoff to realise there are more fixes... From a UX perspective those form labels shouldn't be camelCase. Also the customer says they don't like those form fields being that long. Need to be a bit shorter. Hope this helps :)

File: validator/roleValidator.ts [hunk 5]
```diff
@@ -1,13 +1,35 @@
...
+    /* eslint-enable */
+
     return null;
 };
+
+
+
+
+
+
```
Comment (jkelso662) [inline validator/roleValidator.ts:35]: Watch out for random unneeded whitespace!

**Review by Criofann — DISMISSED** [review]
Looks good to me !

File: controller/roleController.ts [hunk 6]
```diff
@@ -1,16 +1,46 @@
...
+             await roleService.createRole(role);
+        }
+        catch(e) {
+            res.locals.errormessage = e.message;
+            res.render("add-job-role", req.body);
+        } 
+    });
+       
+
+    app.get("/job-role-spec/:roleName", async (req: Request, res: Response) => {
```
Comment (steveh2707) [inline controller/roleController.ts:36]: remove individual role from pull request

File: test/unit/service/roleServiceTests.ts
Comment (steveh2707) [inline test/unit/service/roleServiceTests.ts:1]: why is there a roleService test file and a roleServiceTests test file?

**Pull request merged**